src/components/chatbox/message.tsx

```diff
@@ -0,0 +1,251 @@
+
+// https://helios.hashicorp.design/icons/library
+
+import { IconCheckSquare24 } from '@hashicorp/flight-icons/svg-react/check-square-24'
+import { IconClipboard24 } from '@hashicorp/flight-icons/svg-react/clipboard-24'
+import { IconThumbsDown24 } from '@hashicorp/flight-icons/svg-react/thumbs-down-24'
+import { IconThumbsUp24 } from '@hashicorp/flight-icons/svg-react/thumbs-up-24'
+import { IconWand24 } from '@hashicorp/flight-icons/svg-react/wand-24'
+
+import Button from 'components/button'
+import Text from 'components/text'
+import FeedbackForm from 'components/feedback-form'
+import IconTile from 'components/icon-tile'
+import useAuthentication from 'hooks/use-authentication'
+
+import s from './message.module.css'
+
+export const UserMessage = ({
+	text,
+	image,
+}: {
+	image?: string
+	text: string
+}) => {
+	return (
+		<div className={cn(s.message, s.message_user)}>
+			<div className={cn(s.message_avatar)}>
+				{image ? <img src={image} alt="user avatar" /> : null}
+			</div>
+			<Text /* Body/200/Medium */
+				size={200}
+				weight="medium"
+				className={cn(s.message_content, s.message_user_input)}
+			>
+				{text}
+			</Text>
+			<div className={cn(s.message_gutter)}></div>
+		</div>
+	)
+}
+
+export const AssistantMessage = ({
+	markdown,
+	showActions,
+	conversationId,
+	messageId,
+}: {
+	markdown: string
+	showActions: boolean
+	conversationId: string
+	messageId: string
+}) => {
+	const { session } = useAuthentication()
+	const accessToken = session?.accessToken
+	// Determines green/red button
+	const [rating, setRating] = useState<1 | -1 | 0>(0)
+
+	const handleFeedback = async ({
+		rating,
+		text,
+	}: {
+		rating?: -1 | 1
+		text?: string
+	}) => {
+		const response = await fetch('/api/chat/feedback', {
+			method: 'POST',
+			headers: {
+				'Content-Type': 'application/json',
+				Authorization: `Bearer ${accessToken}`,
+			},
+			body: JSON.stringify({
+				messageId,
+				conversationId,
+				rating,
+				text,
+			}),
+		})
+		return response
+	}
+
+	const [isCopied, setIsCopied] = useState(false)
+	useEffect(() => {
+		let id: ReturnType<typeof setTimeout>
+		if (isCopied) {
+			id = setTimeout(() => {
+				setIsCopied(false)
+			}, 4000)
+		}
+		return () => {
+			clearTimeout(id)
+		}
+	}, [isCopied])
+
+	/** copy rich text; not raw markdown */
+	const handleCopy = () => {
+		// CSS Module classname is a reasonable "hook" to grab the element
+		// rendered by ReactMarkdown.
+		const el = document.getElementsByClassName(s.message_markdown)?.[0]
+
+		if (el) {
+			// https://www.nikouusitalo.com/blog/why-isnt-clipboard-write-copying-my-richtext-html/
+			const clipboardItem = new ClipboardItem({
+				// @ts-expect-error - innerText is available
+				'text/plain': new Blob([el.innerText], { type: 'text/plain' }),
+				'text/html': new Blob([el.outerHTML], { type: 'text/html' }),
+			})
+
+			navigator.clipboard.write([clipboardItem])
```

Comment:
This can (and often does) fail. It returns a Promise, so we should only call `setIsCopied` if the Promise resolves. At this point I wouldn't worry about communicating a failure.

Reply:
Oh TIL! Addressed 7f41e22ed
- handleCopy is now async
- the caught error is ignored